[둔덩] API 테스트/문서자동화 미션 제출합니다.

DunDung opened the pull request:
안녕하세요 !
잘 부탁드립니다 :)

fucct committed:
feat: 실습1, 2 구현

- 문서화 실습 구현
- 인증 실습 구현

fucct committed:
refactor: MemberResponse 기본 생성자 추가

fucct committed:
feat: MethodArgumentResolver에서 password와 password-check가 일치하는 지 검증

fucct committed:
feat: interceptor 와 methodArgumentResolver 에서 발생하는 예외 관리

- 아직 미완성..

fucct committed:
feat: annotation 기반 method argument validate 추가

- 패스워드 일치
- 이메일 중복

fucct committed:
feat: create 예외에 대한 문서화 구현

fucct committed:
feat: get에 대한 서비스, 컨트롤러 테스트와 문서화 구현

fucct committed:
feat: update에 대한 서비스, 컨트롤러 테스트와 문서화 구현

fucct committed:
feat: delete에 대한 서비스, 컨트롤러 테스트와 문서화 구현

fucct committed:
feat: 회원 가입 프론트 연동 구현

fucct committed:
feat: 로그인, 회원 정보 보기 기능 구현

fucct committed:
feat: 업데이트 구현

fucct committed:
feat: 탈퇴 구현

fucct committed:
refactor, feat: 테스트 정상화 및, 문서화 구현

- 전체 테스트 정상화
- 문서화 추가 양식 구현 (필수값 여부, 입력값 foramt)
- 예외 사항 추가

fucct committed:
feat: IsAuth 어노테이션 추가 및 테스트와 문서 수정

fucct committed:
feat: 컨트롤러 테스트 추가

- 토큰이 없는 경우 정보조회 테스트 추가

fucct committed:
feat: 즐겨찾기 서비스 테스트 추가

fucct committed:
feat: 즐겨찾기 관련 도메인 테스트 추가

fucct committed:
feat: 즐겨찾기 관련 도메인 테스트 추가

- 즐겨찾기 삭제 테스트 및 구현

fucct committed:
feat: 즐겨 찾기 프론트 연동 구현

fucct committed:
refactor: 테스트 정상화

jihan805 commented on src/main/java/wooteco/subway/web/member/controller/LoginMemberController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @PostMapping("/login")
+    public ResponseEntity<TokenResponse> login(@RequestBody LoginRequest param) {
```
전체적으로 테스트를 엄청 꼼꼼하게 잘 작성하셨네요 💯 
토큰 생성에 실패하는 경우에 대한 테스트도 추가해보면 어떨까요?

DunDung replied:
추가해보았습니다!

리뷰어님께서는 테스트를 얼만큼 꼼꼼히 해야한다고 생각하시는지 궁금합니다!
테스트결과가 뻔하거나 변하지 않을 것 같은 부분들도 전부 테스트해야할까요??
문서화를 하기 위해서라도 생각가능한 모든 예외를 테스트해야할까요?

jihan805 replied:
작성된 코드와(레벨 1 연습) 추가로 예외가 발생할 수 있는 부분(예외 케이스)에 대해서 의도한 예외를 잘 뱉어내는지 테스트 해주는 편입니다!
작성한 코드 기준으로 멤버 업데이트시, 패스워드 관련 에러 발생 등

jihan805 commented on src/main/java/wooteco/subway/web/member/controller/LoginMemberController.java:
```diff
@@ -0,0 +1,58 @@
+
+    @IsAuth(isAuth = Auth.AUTH)
+    @GetMapping("/me/favorites")
+    public ResponseEntity<Set<FavoriteDetailResponse>> getMemberFavorites(@LoginMember Member member) {
```
Favorite controller 에서 처리하면 어떨가요?
/favorites/me

DunDung replied:
FavoriteController에서 처리하도록 수정했습니다 :)

jihan805 commented on src/main/java/wooteco/subway/web/favorite/FavoriteController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @IsAuth(isAuth = Auth.AUTH)
+    @PostMapping("/favorites")
```
/favorites가 중복되는데 @RequestMapping을 사용하면 어떨까요?

DunDung replied:
@RequestMapping을 사용하도록 수정했습니다!

jihan805 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -32,20 +41,22 @@ public void deleteMember(Long id) {
 
     public String createToken(LoginRequest param) {
-        Member member = memberRepository.findByEmail(param.getEmail()).orElseThrow(RuntimeException::new);
+        Member member = memberRepository.findByEmail(param.getEmail())
```
아래 메서드를 사용할 수 있을 것 같아요!

DunDung replied:
어이쿠... 실수했습니다!
수정했습니다 :)

jihan805 commented on src/test/java/wooteco/subway/domain/member/MemberTest.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @Test
+    void addFavorite() {
```
경로 중복에 대한 케이스도 추가해보면 어떨까요?

DunDung replied:
경로 중복에 대한 테스트 추가했습니다!

jihan805 commented on src/main/java/wooteco/subway/service/member/dto/MemberRequest.java:
```diff
@@ -1,12 +1,44 @@
+    @NotBlank(message = "이름은 필수 입력 사항입니다.")
     private String name;
+
+    @NotBlank(message = "이름은 필수 입력 사항입니다.")
```
이름이 아닌 password가 아닐까요?

DunDung replied:
ㅠㅠㅠ복붙을 하더라도 신경썼어야했는데 ㅠㅠ 수정했습니다!

jihan805 reviewed (changes requested):
안녕하세요, 구현 잘하셨어요 💯 
간단한 코멘트 남겼으니 확인해주세요 :)

DunDung committed:
test: 로그인 실패 테스트 추가

* 이메일이 존재하지 않을 경우
* 패스워드가 일치하지 않을 경우

DunDung committed:
refactor: 리뷰 반영

 - FavoriteController, LineController, MemberController에 @RequestMapping 추가
 - 경로 중복에 대한 테스트 추가
 - @NotBlank 메시지 수정

jihan805 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -1,29 +1,38 @@
-        member.update(param.getName(), param.getPassword());
+        Member member = memberRepository.findById(id)
+            .orElseThrow(() -> new NotFoundMemberException(param.getName()));
+        if (!member.checkPassword(param.getOldPassword())) {
```
중복 코드는 메서드로 추출하면 어떨까요?

jihan805 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -1,14 +1,21 @@
     private String email;
     private String name;
     private String password;
+    private Set<Favorite> favorites = new LinkedHashSet<>();
```
일급 콜렉션을 사용할 수 있을 것 같아요!

DunDung replied:
일급 콜렉션을 사용하도록 수정했습니다 !

jihan805 commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,65 @@
+
+    private Favorite getFavorite(FavoriteRequest request) {
+        Long sourceId = stationService.findStationIdByName(request.getSourceName());
+        Long targetId = stationService.findStationIdByName(request.getTargetName());
```
동일한 쿼리를 두번 날리는 방법도 있지만, 한번에(ids) 조회하여 그 결과를 mapping 해주면 어떨까요?(효율성)

DunDung replied:
`findIdsByNames`라는 메서드를 만들어서 한 번에 조회하도록 수정했습니다 !

jihan805 commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -22,14 +28,17 @@ public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvi
+                String bearer = authExtractor.extract(request);
+                jwtTokenProvider.validateToken(bearer);
+                String email = jwtTokenProvider.getSubject(bearer);
+                request.setAttribute("loginMemberEmail", email);
```
"loginMemberEmail" -> 상수

DunDung replied:
상수로 수정했습니다!

jihan805 reviewed (changes requested):
피드백 반영 잘하셨어요 💯 
간단한 코멘트 추가하였습니다! 확인 부탁드려요 :)

DunDung committed:
refactor: 리뷰 반영

- Member에서 Favorites 일급콜렉션 사용하도록 수정
- Favorite 테스트 추가
- MemberService 예외상황 테스트 추가
- getFavorite 쿼리를 한 번만 날리도록 수정
- import 순서 정리

DunDung committed:
refactor: 리뷰 반영2

- email의 attribute name 상수화

DunDung committed:
style: 코드 스타일 정리

jihan805 reviewed (approved):
안녕하세요~!! 피드백 반영 잘하셨습니다 💯  (구현도!!)
머지할게요 :)
